test/cql-pytest/conftest.py

```diff
@@ -221,3 +226,20 @@ def temp_workdir():
     """ Creates a temporary work directory, for the scope of a single test. """
     with tempfile.TemporaryDirectory() as workdir:
         yield workdir
+
+
+@pytest.fixture(scope="function")
+def xfail_tablets(test_keyspace):
+    if test_keyspace.endswith('_tablets'):
+        pytest.xfail("test fails with tablets")
+
+# Automatically parameterize each test, using test_keyspace (directly or
+# indirectly) with two test_keyspace instances: one which uses tablets, another
+# which uses vnodes.
+# The test_keyspace receives the automatic param, via request.param.
+def pytest_generate_tests(metafunc):
+    if 'test_keyspace' in metafunc.fixturenames:
+        metafunc.parametrize(
+                "test_keyspace",
+                ["tablets", "vnodes"],
+                indirect=True)
```

Comment:
This is a very cool trick. I still think it should NOT be the default to run all the tests twice (and as I said, a year from now it can become 2^N times), at least not without a special parameter. But since the code is now so NOT ugly, I'm less happy to fight against it.

But can you please at least give some thought to the question of when we'll REMOVE this double-testing code? Will we still continue to run each test twice even a year from now, when tablets will have been the default for a long time - but still some users still use vnodes so we can defend the need to continue testing it?

Please also note that in 99% (just a made-up statistic) of cql-pytest tests, tablets do not matter at all. Tablets do not change the correctness of CQL filtering, or TIMESTAMP formatting, or hundreds of other cql-pytest tests. I think we can even decide right now to test **only** with tablets, and stop testing with vnodes.

I wonder if we could we test everything with tablets except the xfail_tablet tests (whose number should go down hopefully soon), which will be tested without tablets?

Reply:
I think that functional tests (cql-pytest) and integration tests (dtest) will run with both with and without tablets until we retire vnode support for user tables. Until such time that we support users opting out of tablets, we cannot afford to have virtually no test coverage for them. AFAIK, we have 0 tests that specifically test that the vnodes algorithm works. Instead, it is tested indirectly by all the different functional and integration tests that happen to write and read from tables.

Reply:
> I think that functional tests (cql-pytest) and integration tests (dtest) will run with both with and without tablets until we retire vnode support for user tables.

Was the same true for "raft-based schema changes"? If not, why is this case different?

> Until such time that we support users opting out of tablets, we cannot afford to have virtually no test coverage for them. AFAIK, we have 0 tests that specifically test that the vnodes algorithm works. Instead, it is tested indirectly by all the different functional and integration tests that happen to write and read from tables.

It doesn't mean that developers need to run these tests twice every time they change the way that CQL filtering works. A different possibility is to add an option, and Jenkins but NOT most developers - will choose to run the tests twice.

Since I think in the context of the vnode-tablet differences it is actually dtests which are more important, you should first think what will be approach in dtest and then decide what this means for cql-pytest. If dtest will need to be run twice by Jenkins, then it can also do this for test.py. (and also note that you change only applies to cql-pytest - what about test/cql? test/alternator? test/toplogy? Why don't we need to run all of those in the same way?).

Personally, I think that test.py (and dtest) should have a tablets/vnode option, which affects the way they run Scylla (and should set a **default** initial_tablets and not require the test code to set it!), and Jenkins would run test.py and dtest twice - once with tablets and once with vnodes. The same option can be in `test/cql-pytest/run`, but I don't think it should run all tests twice.

Reply:
I still don't understand why we need the "--all" option, or the complications that come with it (even the renaming of the test keyspace isn't necessary without it). Why can't test.py just run the test suite twice? Won't you need this ability for other test suites as well? E.g., the C++ unit tests, how will you run those with and without tablets without just running them twice?

Reply:
I think running the suite twice would make the failure reporting and the investigation experience even worse then it is today. With the current approach, the name of the test tells you which variant failed -- vnodes or tablets. If we ran the suites twice, you would have to go hunting in the logs to find out which variant it is.

Also, I want to avoid touching the immensely complex `test.py` if possible. Any changes there will impact not just this, but other test suites too.

As for the other suites, we will see when we get there. I don't think we want to run the C++ suites with/without tablets. Those are unit tests (for the most part), which don't care what replication scylla uses. There are some functional tests in `test/boost/cql_query_test.cc`, we should just migrate those to `test/cql-pytest`, one day.

Reply:
I agree with Avi - as I noted above with a made-up statistic, 99% of the cql-pytest don't care about tablets vs vnodes - it's just one node, and then the test checks some small CQL feature if it returns the right thing or not. It will be very annoying and frustrating to see each of these tests run twice every single time. Note that I don't care if they run 2 or 6 or 74 times during the CI (today they already run between 3 and 3*3 times) - I just don't want developers to have to have to do it.

Reply:
And watch out not to use the fact that test.py is overly-complex as a reason to complicate cql-pytest. Yes, I agree with you that test.py is too complex, andand I've been saying this all along, but it doesn't mean I want to bring all that unnecessary complexity into cql-pytest.

Reply:
> Doesn't this apply to cql-pytest too? Does bug that cql-pytest tests is likely to be triggered by tablets?

Most of the tests in test/boost do not ever run a single query. They instantiate some object and call its API with different parameters and check the results. Tablets/vnodes have no bearing on these kinds of tests..
Like I said, the exception is `test/boost/cql_query_test.cc`, which *does* run queries.

Contrary to that, *all* tests in cql-pytest run queries, and those will all exercise the core replication paths.